umbraco/Umbraco-CMS
# Resolves issues related to the setting of a schedule for content publishing

### Prerequisites


Addresses issues raised in #17031, tracked under [AB 45476](https://dev.azure.com/umbraco/D-Team%20Tracker/_workitems/edit/45476) (internal HQ tracker)

Also fixes #17830 tracked under [AB 47430](https://dev.azure.com/umbraco/D-Team%20Tracker/_workitems/edit/47430)

Also fixes #18297 tracked under [AB 48986](https://dev.azure.com/umbraco/D-Team%20Tracker/_workitems/edit/48986)

### Description

**Reported issue:**

https://github.com/umbraco/Umbraco-CMS/issues/17031 describes several issues related to the setting and reading of scheduled publishing details:

1. It's not possible to see if a page is scheduled in the "info" section of the page.
2. You cannot schedule an item that has not been saved yet.
3. If you try to publish an item that has been scheduled you will get an error:
4. If you click on "schedule" on an already scheduled page the dates for the previous schedule are not displayed.
5. There is no conversion between local and server time on save and display of the scheduled dates.
6. A previously set scheduled date can't be cleared.

**Resolution:**

With the PR merged, these issues will be addressed as follows:

1. In Umbraco 13 the details of the schedule were displayed in the "Info" tab when editing content.  I've added that back now:

<img src="https://github.com/user-attachments/assets/91416b97-cdcb-48c3-b9f8-498a504b468b" width="250" />

2. Conditions have been added now so that all the actions available as workspace action button extensions - unpublish, publish with descendants and schedule - no longer appear until the content has been saved for the first time.  They all error if called on a new content item that hasn't yet been saved.
3. Getting an error presented when trying to publish and item that has been scheduled is correct behaviour, or at least aligns with what we have on 13, so I've made no changes to this.
4. Previously selected schedule is now displayed when returning to the dialog to view or change the schedule.

<img src="https://github.com/user-attachments/assets/038ca098-522f-406a-936b-2149b4442195" width="450" />

5. A conversion is now made between the local and server time when saving and presenting the schedule.  There's no conversion when running locally as of course the server and local time are the same, but this update should resolve issues where the times aren't the same, and aligns with the behaviour we had on 13.
6. A previously set schedule can now be cleared.

**To Test:**

- Use the schedule dialog for a new and existing content item to verify issues 1, 2, 4 and 6 are resolved.
- Go to DevTools in your browser and enable the **Sensors** menu, set your timezone to something else than the server and schedule a document. Verify that the saved schedule is saved to the database with the same value as we have in an Umbraco 13 site (in `umbracoContentSchedule`), which will indicate that issue 5 is resolved.
- Change your timezone to something else and verify that the time displayed for scheduling is relative.

### Caveats

1. There is no validation on the times set as of yet (i.e. you can set an unpublish time before the publish time, etc). The server will fail the scheduling, but we can prevent it in the client before submitting. We will do this in another PR.
2. The document is not **saved** before scheduling, which was the behavior in V13. That also means that no validation is being run on the fields allowing you to circumvent the validation system. We will also handle this in another PR.
